ls1intum/Artemis
# `Programming exercises`: Add the option to record the test-wise coverage for Java/Kotlin programming exercises with Maven

### Checklist
#### General
#### Server
#### Client
#### Changes affecting Programming Exercises
- ~[ ] I tested **all** changes and their related features with **all** corresponding user types on Test Server 2 (Jenkins and Gitlab)~. _not applicable_

### Motivation and Context
Closes #4963.

These changes are part of the Bachelor Thesis from @Hialus and @ole-ve about the automatic generation of code hints for programming exercises (HESTIA) in Java.

The option to perform a testwise coverage analysis for the solution repository of Java Gradle programming exercises has been implemented in #4889. The resulting report is needed to generate behavioral solution entries for an exercise, from which code hints are generated. These changes extend this functionality to also work for (Java/Kotlin) Maven exercises. While only the generation of hints for Java exercises is in the scope of our thesis, I decided to also add the testwise coverage report option for Kotlin exercises since this can be done easily.

### Description
The utilized Teamscale JaCoCo Agent for generating the coverage report now also supports the execution for Maven projects.
If the option to perform a testwise coverage analysis has been selected on exercise creation, the `pom.xml` (in the test repository) contains a new profile for the test execution including the coverage analysis. This profile will only be executed for solution submissions, and not for template/student submissions. In case the option is not selected, the generated `pom.xml` is not changed in comparison to the current implementation.

### Steps for Testing
Prerequisites:
- 1 Editor on Test-Server with the Atlassian Stack (TS1, TS3)

1. Create a new programming exercise with the following configuration
- Programming Language: Java
- Project Type: Maven (selecting `with exemplary dependency` is not necessary)
- Activate SCA
- Activate the option "Record testwise coverage"
3. Wait for the solution build plan to finish
4. Click the button `Show testwise coverage` from the programming exercise details page.
5. Verify that the coverage report is displayed as in #4889.
6. Repeat step 1, but use Kotlin instead (note that package names in Kotlin cannot contain "in")

### Review Progress
#### Code Review
#### Manual Tests

### Test Coverage


| Class/File | Branch | Line |
|------------|-------:|-----:|
_no changes_

### Screenshots
For Kotlin exercises:
<img width="1133" alt="Bildschirmfoto 2022-06-08 um 22 05 36" src="https://user-images.githubusercontent.com/73833780/172707989-a98cd2f1-673f-44f6-9301-e3d354e70c5d.png">
